### ISSUE TITLE
Suggest adding "with" statement to C#

### ISSUE THREAD

**Issue by rushfrisby** [dotnet/csharplang#1534]
This would work similar to the [VB With statement](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/with-end-with-statement)

The syntax in C# could look something like:

```csharp
with(person)
{
    .FirstName = "Rush";
    .LastName = "Frisby";
}
```

alternatively it could look like:

```csharp
person.
{
    FirstName = "Rush",
    LastName = "Frisby"
};
```

or both ways!

This seems reasonable seeing as how there is already some behavior like this for object initialization.

**Comment by ufcpp** [comment]
https://blogs.msdn.microsoft.com/csharpfaq/2004/03/11/why-doesnt-c-have-vb-nets-with-operator/
https://github.com/dotnet/csharplang/blob/0709623617764880198f972d66845e88a89fe149/meetings/2015/LDM-2015-03-25-Notes.md

**Comment by yaakov-h** [comment]
Related: #133.

**Comment by bondsbw** [comment]
@ufcpp I may have to \*gasp\* disagree with Jon Skeet on that.

A `with` operator (combined with braces) encourages grouping interactions with the object.  That is clearer than the alternative, and I feel it is one motivation behind today's initializer syntax.

That said, I like its current proposed form as way to "mutate" immutable objects.

**Comment by HaloFour** [comment]
@bondsbw 

It wasn't Jon Skeet's opinion, he was just reblogging [this post](https://web.archive.org/web/20040214165656/http://msdn.microsoft.com/vcsharp/team/language/ask/withstatement/default.aspx) from Scott Wiltamuth.

With `var` and simple named variables I don't find `with` particularly compelling, even if you avoid the massive design cluster-fudge that ended up in EcmaScript.  I even recall proposals in VB.NET to modify `With` to have the ability to "name" the target expression, which is basically the same thing.

**Comment by bondsbw** [comment]
Fair enough, this proposal (and what that link addresses) is a `with` statement.

I was thinking `with` _expression_, which can be more useful in expression-bodied context.

**Comment by rushfrisby** [comment]
Scott Wiltamuth's arguments against `with` were valid-ish back in 2004 since the language was new but not so much any more. With the expectation that variable names should be well-named to minimize comments fragmenting readability this makes even more sense, as well-name variables are usually longer. This is the scenario I which led me to creating this issue. Take the following for example:

```csharp
var updatedCustomAction = MappingHelper.Map<DbCustomAction>(customAction);
updatedCustomAction.CreatedBy = existingCustomAction.CreatedBy;
updatedCustomAction.CreatedOn = existingCustomAction.CreatedOn;
updatedCustomAction.UpdatedOn = DateTimeOffset.UtcNow;
updatedCustomAction.UpdatedBy = currentUserId;
updatedCustomAction.DeletedBy = null;
updatedCustomAction.DeletedOn = null;
```

versus

```csharp
var updatedCustomAction = MappingHelper.Map<DbCustomAction>(customAction);
with(updatedCustomAction)
{
	.CreatedBy = existingCustomAction.CreatedBy;
	.CreatedOn = existingCustomAction.CreatedOn;
	.DeletedBy = null;
	.DeletedOn = null;
	.UpdatedOn = DateTimeOffset.UtcNow;
	.UpdatedBy = currentUserId;
}
```

which is a lot easier to read in my opinion.

**Comment by yaakov-h** [comment]
@bondsbw you're right, I somehow got the wrong issue number.

I meant #803, though #77 and #162 are more closely related to your specific examples, I think.

**Comment by renatoprogramer** [comment]
@rushfrisby 

**Sugestion**
```
var updatedCustomAction = MappingHelper.Map<DbCustomAction>(customAction);
with updatedCustomAction
{
	CreatedBy = existingCustomAction.CreatedBy,
	CreatedOn = existingCustomAction.CreatedOn,
	UpdatedOn = DateTimeOffset.UtcNow,
	UpdatedBy = currentUserId,
	DeletedBy = null,
	DeletedOn = null
}
```

**Comment by WrongBit** [comment]
Before I offered same 'with' operator and now also support rushfrisby's offer. It's so obvious simplification of code that it's surprise to me people didn't implement it yet, but discuss "how to make total mess, introducing 'class ABC (some variables)' rubbish". Above just "votes" C# team should use brains to filter really helpful features. When they made object initializers, THAT was "almost WITH operator", but half-baked due to narrow vision. Now we again and again ask for absolutely simplistic thing - introduce "with" statement which will clean-up boilerplate code.

**Comment by HaloFour** [comment]
@WrongBit 

The C# team hasn't offered `with` because they haven't found it compelling enough to be worth championing through the language design process.  That process is not driven by "votes", particularly not from here.  The language curation process is lengthy with an exceptionally high bar due to the vast cost in implementation and support, the vast majority of proposed language features will fall way short of that threshold.

My opinion is that the `with` feature simply doesn't make the cut in terms of being useful enough to implement.  It doesn't solve any problems nor does it appreciably reduce verbosity.  Whether or not you like it you can easily achieve the same results by simply assigning an expression to a temporary variable with as terse of a name as you'd like.  I think that any language feature would have to demonstrate that it offers more than trading a few characters in an identifier with a few characters in an indentation.

**Comment by WrongBit** [comment]
@HaloFour You're talking like last 10 years you slept. Sorry, but today we cannot afford to waste time on boilerplate code! By your logic every single "syntactic sugar" can be easy removed, because "doesn't worth championing". But I insist - anything what reduces repetitive/boring typing should be implemented. Esp. when it's not "unknown, doubtful feature", but operator from existing language with predictable logic and usefulness. I do not care if you personally won't use it, I care to have this feature, because I have enough qualification to see how it helps me.

**Comment by HaloFour** [comment]
@WrongBit 

> Sorry, but today we cannot afford to waste time on boilerplate code! 

> But I insist - anything what reduces repetitive/boring typing should be implemented.

It's not a goal of C# to make the language as terse as possible.  The difference in verbosity between a proposed `with` statement and assigning an expression to a single-character identifier is negligible at best.

Even if this proposal was less verbose by a noticeable degree, that's still not enough of a reason for it to be added to the language.  It has to fit in to the design goals of the language design team and be deemed worth the massive amount of time and effort to implement and support for perpetuity.

> Esp. when it's not "unknown, doubtful feature", but operator from existing language with predictable logic and usefulness.

You're free to use those languages.  [MS even ships one.](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/statements/with-end-with-statement).

**Comment by WrongBit** [comment]
OK, I got it - MS do not give a damn about any our requests: they have **their own** "language goal" which nobody knows. That's why request of **real developers** on C# have nothing common with features C# team implements. Sorry, but again about qualification: you write too few code in real applications to discuss what "worth" or not. I do. Every single day since .NET 1.1; So **I** deal with every day verbosity of C#, not you. That's why you completely don't see usefulness of "with" and produce pathetic excuses why it should not be in the language. It's ridiculous - people who just write C# compiler rule the people who use it.

**Comment by YairHalberstadt** [comment]
@wrongbit

First you state:

> Above just "votes" C# team should use brains to filter really helpful features.

The you state

> That's why request of *real developers* on C# have nothing common with features C# team implements.

These statements seem to be in contradiction to each other. First you suggest the C# designers should not listen to what developers ask for and should make their own decision, then you suggest that they should not make their own decisions but should do what people ask them. Which is it?

Regardless this request has neither votes, nor backing by the LDM, so either way is unlikely to happen.

**Comment by YairHalberstadt** [comment]
> It's ridiculous - people who just write C# compiler rule the people who use it.

The compiler is written in C#, so the people who write it also use it.

**Comment by orthoxerox** [comment]
@WrongBit 

MSFT do have their own language goals that are shared with the broader community via Build conferences, Channel 9 and this repo.

It's one thing to disagree with the direction that are taking the language in, but it's a whole other thing to claim they do not listen to external input. C# is one of the top 10 most used languages, so it looks like it fits the needs of people reasonably well.

MSFT absolutely does solicit feedback on the direction of C#. The listen to ISVs writing programs in C#, they listen to companies with large internal C# shops, they listen to C# MVPs, they listen to the broader community.

Yes, there are aspects of C# that do not serve the needs of LOB application programmers reasonably well. I can name:

- property-scoped backing fields
- decorated properties (think auto-implementing `INotifyPropertyChanged`)
- and yes, `with` statements

as feature requests that don't get enough attention from the design team compared to the amount of discussion they generate.

We don't have a confirmed answer why, but I can suggest several ideas:

- there are more groundbreaking features they feel they need to work on
- they do not want these features because they think they push C# programmers in the wrong direction (for example, by making writing large mutable objects easier)
- they do not want to add new features to C# unless there's a universal demand for them (I've seen people online complain that C# already has too many features, comparing it to Perl and Scala)

**Comment by HaloFour** [comment]
@orthoxerox 

Of those three mentioned aspects I'd suggest that at least the decorated properties is on their radar, but they're hoping for a more general purpose solution.  Hopefully source generators gets more attention shortly and that will achieve those goals.

I disagree that `with` really brings anything to the table.  Whereas implementation of `INotifyPropertyChanged` (and the dozens of other similar use cases) can significantly reduce boilerplate, `with` doesn't reduce code by any substantial amount.  So you trade:

```cs
var a = Some.Expression.Here;
a.Property = foo;
a.DoSometing();
var x = a.SomeOtherProperty;
```

for:

```cs
SomeType x;
with (Some.Expression.Here) {
    .Property = foo;
    .DoSomething();
    x = .SomeOtherProperty;
}
```

Is there really a benefit here?  I don't think so.  It's not less verbose and it introduces other caveats like scoping that will require developers to have to work around.  Even if you ignore those cases you'd probably save at most a handful of characters if you happen to have a large number of interactions within the block.

And given I've seen proposals for VB.NET (I think on CodePlex, I failed to locate them when writing this) where people are requesting another magickal keyword that would let you gain access to the expression from the `With` statement so that it could be passed to a method or stuff like that it just seems even more superfluous.

I loves me some syntax candy, but `with` just feels like copying features from other languages for the sake of doing so.  And I came from a VB/Pascal background and have enjoyed using the feature in those languages.

**Comment by WrongBit** [comment]
@YairHalberstadt You mix things like "vote" and "request". Feel the difference and read my comment again.

**Comment by YairHalberstadt** [comment]
@WrongBit

> You mix things like "vote" and "request". Feel the difference and read my comment again

A vote is a request to microsoft to implement a feature. Nothing more, nothing less.

> Again you mix use of C# for "compiler writing" with MUCH WIDER usage of C# by all our community. Databases, 3D, games, networking... we do many different things. That's why we ask so different features.
But in any case "with" will be useful for all of us.

I'm not a member of the compiler team. I don't work for Microsoft. Neither does @HaloFour. Neither of us feels with will be useful. In all the cases where it has come it has been very divisive. Many people want it, many don't. Let's accept that whilst you, and many others do think this will make your life easier, you don't speak for everyone.

**Comment by WrongBit** [comment]
@HaloFour Your example just ridiculous and you really think it proof your words? See this one:
````
var veryGoddamnLongMeaningfulName = GetObject();
veryGoddamnLongMeaningfulName.Prop1 = "...";
veryGoddamnLongMeaningfulName.Prop2 = "...";
veryGoddamnLongMeaningfulName.Prop3 = "...";
veryGoddamnLongMeaningfulName.Prop4 = "...";
````
Using "with" we reduce UNNECESSARY code to read:
````
with(var veryGoddamnLongMeaningfulName = GetObject()) {
    .Prop1 = "...";
    .Prop2 = "...";
    .Prop3 = "...";
    .Prop4 = "...";
}
````
Only blind people cannot see how SIMPLER code became. Blind in all meanings. I'm so upset I need to explain childish thing to people who pretend to be developers.

**Comment by spydacarnage** [comment]
@WrongBit  - In all your time here on this repo, have you ever found that insulting people is the way to get what you want?

(pro tip - it isn't)

**Comment by WrongBit** [comment]
@HaloFour Related "INotifyPropertyChanged" - just trust me, 10 years later people will kill you for any mention of that ugliest "programmer's invention" (like now for 'goto'). Sorry for offtop.

**Comment by YairHalberstadt** [comment]
@WrongBit

> Only blind people cannot see how SIMPLER code became. Blind in all meanings. I'm so upset I need to explain childish thing to people who pretend to be developers.

Can you please try and avoid being impolite when discussing things here. It is making this conversation unpleasant.

**Comment by WrongBit** [comment]
@YairHalberstadt First I need to see who discuss "with" feature. ATM I see nobody who shows at least one serious argument against. So yes, if you feel unpleasant having no arguments, just leave conversation.

**Comment by DanFTRX** [comment]
This is just rehashing this conversation:  https://github.com/dotnet/roslyn/issues/6553

**Comment by mburbea** [comment]
@WrongBit, I don't see why this can't be solved with some previously suggested library method to let you give it a nice scoped short name.
```cs
public static void With<T>(this T foo, Action<T> action) => action(foo);
no need to yell at people.
Of course, you could always just do that inline...
```
{
     var s = SomeAwfulEldritchUnspeakablyLongName;
     s.Prop1 = "";
       ... 
     s.PropN = "";
}
```
I don't think this situation is very common. Usually, I either initialize and modify after newing up, and in the rare cases I can't because it is given to me by IoC, I rather fix those cases by configuration rather than in code.

**Comment by CyrusNajmabadi** [comment]
>  Blind in all meanings. I'm so upset I need to explain childish thing to people who pretend to be developers.

To give you some context.  I was a member of the C# language design group for over a decade.  I helped create C# versions 2 through 8.  So i've got a lot of experience and thoughts about this.

First off, your attitude and communication style aren't doing you any favors.  At best, they are turning people just not want to listen to you. At worst, it is a violation of the rules of conduct of github and the dotnet foundation.  You need to change this, or you will get banned from participating here.

Second, ignoring your communication style and just focusing on your opinions:

> Only blind people cannot see how SIMPLER code became. 

Simplicity is often in the eye of the beholder.  One person's simplicity is another's complexity. See, for example, how the C++ community views simplicity, and how that differs from the Go community's views.  It is erroneous to assume that your own perspective on what is simple is universally true across the ecosystem and community that C# targets.  

For example, some people find things simpler the *less* they need to look around to pull in necessary context.  With `with` you now have another piece of context you need to understand when examining code.  That brings a form of simplicity by allowing you elide, but it adds complexity in makin people having to understand what that elision means. 

For example, say i have this:

```
    this.very.long.statemnet.etc.etc.etc()
    .method();
```

versus

```
    this.very.long.statemnet.etc.etc.etc();
    .method();
```

That's the difference of one token, but it vastly changes the meaning of thins.  Previously i could look at the first one and know that `.method` must extend the previous line.  In the new world it could mean two things, and my life is more complex because i have to go figure that out.

--

Finally, please consider that some people here have been doing this a long time, and have a lot of experience trying to figure out the particular pros/cons of a proposal and if they would be net positive for the language.  Many will disagree with you.  If you want conversations to proceed toward your desired ends you would do well to work *with* those that disagree, instead of being hostile to them and lashing out.

Cheers, and good luck!

**Comment by HaloFour** [comment]
@WrongBit 

> Your example just ridiculous and you really think it proof your words?

If the name of the identifier is so important that you feel that you need to give it a "long meaningful name" then you'd likely not be using a feature like `with` that gives you no name at all.  The beauty of a temporary variable is that you get to make the choice as to how descriptive or terse you want to be.

I don't understand your example of assigning a variable within the `with` statement.  Is that to address my comment about other proposals wanting to actually access the value of the expression from within the block?  If so, what's the point of the `with` statement at all?  Just use a normal block and a temporary variable declared within that block.

> ATM I see nobody who shows at least one serious argument against.

Ignoring the arguments you don't like doesn't mean that they don't exist.  But more critically, you'd need serious arguments *for* the proposal before the language team would consider adding such a feature.  They've clearly known about `with` since before C# 1.0 was released so it's not like these proposals are showing them something that they don't know about.

As for the tone, I believe that I've been civil towards you even if I've been opinionated towards the feature.  I'd appreciate responses in kind.  If a rebuttal of your argument can only be seen as "childish" there's little chance that we can actually have a meaningful conversation here.

**Comment by WrongBit** [comment]
@CyrusNajmabadi 
I'm happy somebody experienced come to the thread, but again I see nothing serious why "with" should not be in the language. Your example with "it just)[keeps].going;" is kind of "If you put monkey on the keyboard, your code won't compile". Since when "typo" become argument against features? It's a LANGUAGE, it's IT - serious thing where one character can change everything. Say, "if (a)" against "if (!a)". Why nobody of you ("language design group for over a decade") didn't remove "!" sign? :) Now do you feel how weak your argument? And if you still sure somebody said anything meaningful, just cite that guy. I promise, I carefully consider any argument except "our lazy team just don't wanna bore with that".

**Comment by WrongBit** [comment]
@HaloFour  

> The beauty of a temporary variable is that...

Tell me, are you seriously talking now? So I must dirt my code with "temporary variable" just to remove necessarity to read long names? Well, no words... I'm done. I feel it's way easier to make own language than to explain others how code readability is important.

Anybody of you have any guess why people like "C"? One of my args was "I don't have to read those verbose 'begin...end' and can simply put { }". We fight with verbosity for years, even C# team guessed to make "short properties" like ``int a => b;`` instead of ``int a { get { return b; } }``. Why you did that? Was it so hard to put more braces? :) So my idea the same - easier code to read, more errors you avoid.

**Comment by WrongBit** [comment]
Thanks to @DanFTRX he found my old request from 2015, because @rushfrisby miss one important thing we can do, having "with" - adding delegate! My sample:
````
with(var MyVeryDescriptiveVariable = new SomeClass(), OtherVar) {
    .Field1 = .Field2;// it was MyVeryDescriptiveVariable.Field1 = OtherVar.Field2
    .Event += meth();// meth() w/o dot means method from outer scope
    MyVeryDescriptiveVariable.Fun();// ambiguous methods are called with full name
}
````
Funny, but "serious guys" miss this point with ".Event" and we got half-baked "object initializer", which do NOT allow assign event! My proposal (https://github.com/dotnet/roslyn/issues/6553) fixes this mistake + adds ability to easy assign properties not only after object creation (again limitation!), but anywhere in a program.

**Comment by CyrusNajmabadi** [comment]
> but again I see nothing serious why "with" should not be in the language

I was 100% serious with my critique.  

> Your example with "it just)[keeps].going;" is kind of "If you put monkey on the keyboard, your code won't compile". Since when "typo" become argument against features? 

It's been an argument since v1.  We think strongly about these scenarios and the impact it has on customers.  Furthermore, the LDM has the absolute final say here.  So you need to operate within the perspectives and considerations that hte LDM members have here.

I laid out a trivial example of how this language feature makes things more complex.  I now cannot trivially tell what `.foo()` means anymore.  Previously it *always* meant `a invocation off of the previous expression`.  Now it means `an invocation off of the previous expression, if there is one, otherwise it it an invocation off of the innermost contained "with'ed" expression`.  That is *by definition* more complex.

i brought this up not to indicate that the feature shuld not be be done, but to help you with *perspective*.  **Your** perspective is that this is strictly simpler.  However, i showed a completely reasonable perspective that demonstrates the additional complexity added.  You need to be able to see things through a multitude of **perspectives** if you want to be able to convince others of the value of these features.

>  I promise, I carefully consider any argument except "our lazy team just don't wanna bore with that".

Let me make this as clear and as blunt as possible.  I decided to reach out to try to help you here.  If you're going to behave in a childish and petulant manner, then i'll simply walk away as i have no interest engaging in people who behave this way.  You will then lose someone who could have been an advocate for your position.

So ask yourself this:

"What am i trying to accomplish?"

If your goal is to get this language proposal accepted, you are burning bridges and doing a great job to turn people against it entirely based *solely* on your behavior.

However, if on hte other hand, your objective is to feel good about yourself and think you can browbeat people into supporting you, you can continue in this fashion.  But i can assure you, there has been a 0% success rate with people doing this in the past.

So what are you trying to accomplish?  And how will you change your behavior to have a better chance to get what you want.

--

To help you out, i'm willing to respond *one* more time.  It will be based on the behavior and attitude of your next posts.  If they are rude and childish, I'll let you know and simply walk away.  If they demonstrate a willingness to engage and work with others in a civil and adult fashion, I'll be happy to keep communicating and sharing my several decades of experience working directly in this space.  It's up to you how you proceed.

**Comment by mburbea** [comment]
Let's assume you did have a With statement. Cyrus's example is interesting as he shows a potential error vector. Let's simplify it
```cs
with(f){
     .Prop = SomeChained()
     .Calls();               
}
```
now if I accidentally introduce a semicolon
```cs
with(f){
      .Prop = SomeChained();
      .Calls();
}
```
These segments may both be valid C#. In the first we are assigning `Prop` to `SomeChained().Calls()`, but in the later, we are assigning `Prop` to `SomeChained()` and invoking `Calls()`. This can have wildly differing meaning.
This is **alarming** and a subtle bug that is hard to spot.

This reminds me of the danger of semicolon completion in javascript, and this would immediately make me not want to use this method at all.

The question when it comes to language feature is not "Why Not?". It needs to be shown as to why it is a large enough value add to be worthwhile. Everyone writes properties, and trimming down the noise to `public int Foo=> 7;` was considered worthwhile. I can't think of a situation where I would use this beyond the simple initialization (which new T(){...} already handles).

**Comment by HaloFour** [comment]
@WrongBit 

> So I must dirt my code with "temporary variable" just to remove necessarity to read long names?

Temporary variables are not dirt.  They are the idiomatic way to approach a problem such as this in many languages.  They are how `with` is implemented.  Not wanting to use existing features of the language is not a reason to add new features.

**Comment by quinmars** [comment]
> My opinion is that the `with` feature simply doesn't make the cut in terms of being useful enough to implement. It doesn't solve any problems nor does it appreciably reduce verbosity. 

I also don't like the proposed `with` syntax, but I like the second proposed sytax, because it is an expression, which cannot easily mimiced with a temporary variable. And since it is very similar to the initializer syntax, it'd fit very well into C#.

Here a small example. Suppose we have a class that holds the minimum, maximum and the mean value of a data series plus some decriptive properties.

```csharp
public class ValueStatistics
{
    public ValueStatisitics() {}
    public ValueStatisitics(IEnumerable<double> series) { ... }

    public double Minimum { get; set; }
    public double Maximum { get; set; }
    public double Mean { get; set; }
    public string ValueName { get; set; }
    public string Unit { get; set; }
    public string Location { get; set; }
}
````

Now we have a property that returns the statistics for all values in a list:

```csharp
public IEnumerable<ValueStatistics> StatisticallyValues => AllValues
    .GroupBy(v => v.ValueName)
    .Select(g => new ValueStatistics(g.Select(v => v.Value))
    {
        ValueName = g.Key,
        Unit = g.First().Unit
    });
```

It turns out that some values are very off and we don't want to consider those outliers in the statistics. So we write a factory method for `ValueStatistcs` that does exactly this.

```csharp
public IEnumerable<ValueStatistics> StatisticallyValues => AllValues
    .GroupBy(v => v.ValueName)
    .Select(g => 
    {
        var s = ValueStatistics.WithoutOutliers(g.Select(v => v.Value));
        s.ValueName = g.Key;
        s.Unit = g.First().Unit;
        return s;
    });
```

So we have to edit several lines (adding `s.` and replacing `,` by `;`, adding the `return`, etc.) for a rather simple change: replacing a constructor by a factory method.

With the second proposed syntax it'd be only one or two line changes.

```csharp
public IEnumerable<ValueStatistics> StatisticallyValues => AllValues
    .GroupBy(v => v.ValueName)
    .Select(g => ValueStatistics.WithoutOutliers(g.Select(v => v.Value)).
    {
        ValueName = g.Key,
        Unit = g.First().Unit
    });
```

The only problem is that the dot doesn't really fit into the Allman style. On the other hand I have the impression that the `with` keyword is reserved for the _records_ feature.
